### PR TITLE
New account

### PR DESCRIPTION
**Pull request by pablo-vazquez01**
all UI is created. New account created with corresponding unit tests

### CODE REVIEW

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    
+    @IBOutlet weak var passwordInput: UITextField!
+    
+    @IBOutlet weak var passwordConfirmInput: UITextField!
+    
+    @IBOutlet weak var warningLabel: UILabel!
+    
+    @IBOutlet weak var submitButton: UIButton!
+    
+    override func viewDidLoad() {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: If you don't do anything aside from calling super function, please delete.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        super.viewDidLoad()
+
+
+    }
+    
+    
+    @IBAction func clickedSubmit(_ sender: Any) {
+        if (checkInput()){
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Use guard instead for cleaner code.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+//  CreateAccountViewController.swift
+//  PokemonApp
+//
+//  Created by Mac Consultant on 6/9/19.
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import UIKit
+
+class CreateAccountViewController: UIViewController {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Unless a view controller needs to be subclassed, it should be final.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    @IBOutlet weak var usernameInput: UITextField!
+
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Remove these extra lines between the IBOutlets.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift (hunk 2 above)
Comment (prrenfrew) [reply PokemonApp/PokemonApp/CreateAccountViewController.swift]: Also, don't use parentheses around your if checks. There also needs to be a space between the check and the opening brace.

`if someBool {`

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+
+    }
+    
+    
+    @IBAction func clickedSubmit(_ sender: Any) {
+        if (checkInput()){
+            do{
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: There should be a space between do and the opening brace.

`do {`

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                try Credentials.addAcount(username: usernameInput.text ?? "", password: passwordInput.text ?? "")
+
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Remove this extra blank line.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    }
+    
+    
+    @IBAction func clickedSubmit(_ sender: Any) {
+        if (checkInput()){
+            do{
+                try Credentials.addAcount(username: usernameInput.text ?? "", password: passwordInput.text ?? "")
+                
+            }catch let error as KeychainError{
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: There should be spaces between the beginning and end of your catch statement
`} catch let error as KeychainError {`

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+                switch error{
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Brace spacing.

`switch error {`

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+                case .noPassword:
+                    warningLabel.text = "please enter a password"
+                    print("no password was given")
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: For production code, debugging print statements should be removed.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+                    return
+                case .unhandledError(status: -25299):
+                    warningLabel.text = "That username is already being used. Please try again"
+                    return
+                case .unexpectdPasswordData:
+                    print("unexpected Password data")
+                    return
+                case .unhandledError(let status):
+                    print("Error: \(status) was given")
+                    return
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Why are there so many returns?

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+                }
+            }catch{
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Brace spacing.

`} catch {`

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+            warningLabel.text = ""
+            dismiss(animated: true, completion: nil)
+            return
+        }
+        print("account not created fix error and try again")
+      
+    }
+    
+    func checkInput() -> Bool{
+        if (usernameInput.text?.count ?? 0 < 6){
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: You use this count multiple times. Store it in a local variable and use that instead. Also, remove parentheses.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+            warningLabel.text = "Please enter a username"
+            return false
+        }
+        else if (passwordInput.text?.count ?? 0 < 6  || passwordConfirmInput.text?.count ?? 0 < 6){
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Closing if parentheses and next else if should be on same line.
if someBool {
} else if someOtherBool {

Comment (prrenfrew) [reply PokemonApp/PokemonApp/CreateAccountViewController.swift]: Also brace spacing.

File: PokemonApp/PokemonApp/LoginController.swift [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import Foundation
+
+extension Credentials{
+    
+    static func addAcount(username: String, password: String) throws{
+        
+        var credentials = Credentials(username: username, password: password)
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: Why is this a var?

File: PokemonApp/PokemonApp/LoginController.swift [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+//
+//  Created by Mac Consultant on 6/7/19.
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import Foundation
+
+extension Credentials{
+    
+    static func addAcount(username: String, password: String) throws{
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: Brace spacing.

File: PokemonApp/PokemonApp/LoginController.swift [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+        
+        var credentials = Credentials(username: username, password: password)
+        
+        let account = credentials.username
+        let password = credentials.password.data(using: String.Encoding.utf8)!
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: Don't force unwrap.

File: PokemonApp/PokemonApp/LoginController.swift [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+        
+        var query: [String: Any] = [kSecClass as String: kSecClassGenericPassword,
+                                    kSecAttrAccount as String: account,
+                                    kSecValueData as String: password]
+        
+        let status = SecItemAdd(query as CFDictionary, nil)
+        guard status == errSecSuccess else { throw KeychainError.unhandledError(status: status) }
+
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: Maybe there should be some indication that this function was successful?

File: PokemonApp/PokemonApp/UserCredentials.swift [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+//  UserCredentials.swift
+//  PokemonApp
+//
+//  Created by Mac Consultant on 6/7/19.
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import Foundation
+
+struct Credentials{
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/UserCredentials.swift]: Brace spacing.

File: PokemonApp/PokemonApp/UserCredentials.swift [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+    var username: String
+    var password: String
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/UserCredentials.swift]: Should these change?

File: PokemonApp/PokemonApp/UserCredentials.swift [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+    
+    
+    init(username: String, password: String){
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/UserCredentials.swift]: Brace spacing.

File: PokemonApp/PokemonApp/UserCredentials.swift [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+        self.username = username
+        self.password = password
+    }
+
+}
+
+enum KeychainError: Error{
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/UserCredentials.swift]: Put this into its own file.

File: PokemonApp/PokemonApp/LoginController.swift (hunk 15 above)
Comment (prrenfrew) [reply PokemonApp/PokemonApp/LoginController.swift]: Credentials is a model, correct? This would be outside of the role of a model.

File: PokemonApp/PokemonApp/ViewController.swift [hunk 22]
```diff
@@ -9,8 +9,17 @@
 import UIKit
 
 class ViewController: UIViewController {
-
-  override func viewDidLoad() {
+    
+    @IBOutlet weak var usernameTextfield: UITextField!
+
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/ViewController.swift]: Remove extra lines.

File: PokemonApp/PokemonApp/ViewController.swift [hunk 23]
```diff
@@ -9,8 +9,17 @@
...
+    @IBOutlet weak var usernameTextfield: UITextField!
+    
+    @IBOutlet weak var passworldTextfield: UITextField!
+    
+    @IBOutlet weak var loginButton: UIButton!
+    
+    @IBOutlet weak var newAccountButton: UIButton!
+    
+    
+    override func viewDidLoad() {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/ViewController.swift]: This should be lined up with the closing brace.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+//
+//  Created by Mac Consultant on 6/9/19.
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import XCTest
+import Foundation
+@testable import PokemonApp
+
+class PokemonAppTests: XCTestCase {
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: Classes that don't need to be subclassed should be final.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+
+
+    var username = "tester4"
+    var password = "testerspassword3"
+
+
+
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: A lot of extra lines here. Remove them.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+            SecItemCopyMatching(query as CFDictionary, UnsafeMutablePointer($0))
+        }
+
+
+        guard let existingItem = queryResult as? [String : AnyObject],
+            let passwordData = existingItem[kSecValueData as String] as? Data,
+            let returnPassword = String(data: passwordData, encoding: String.Encoding.utf8)
+            else{
+                //if this is entered something is wrong and the test should fail
+                XCTAssertTrue(false)
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: Use XCTFail() instead.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+
+import XCTest
+import Foundation
+@testable import PokemonApp
+
+class PokemonAppTests: XCTestCase {
+
+
+    var username = "tester4"
+    var password = "testerspassword3"
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: It looks like these are only used for one test. I think it would be a bit better to make them local variables as opposed to properties. Or at least these should be private.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+
+
+
+    func testCreateAccount(){
+        try! Credentials.addAcount(username: username, password: password)
+
+        
+       let query = setupQuery()
+        
+
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: Remove extra lines.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+        var queryResult: AnyObject?
+        withUnsafeMutablePointer(to: &queryResult) {
+            SecItemCopyMatching(query as CFDictionary, UnsafeMutablePointer($0))
+        }
+
+
+        guard let existingItem = queryResult as? [String : AnyObject],
+            let passwordData = existingItem[kSecValueData as String] as? Data,
+            let returnPassword = String(data: passwordData, encoding: String.Encoding.utf8)
+            else{
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: Brace spacing. Also in guard statements that span multiple lines, `else {` should be on the same line as the last line of the guard.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+    
+    private func setupQuery() -> [String: AnyObject]{
+        
+        var query = [String : AnyObject]()
+        
+        query[kSecClass as String] = kSecClassGenericPassword
+        
+        query[kSecAttrAccount as String] = username as AnyObject?
+        
+
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: Remove extra lines.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 31]
```diff
@@ -0,0 +1,57 @@
...
+        XCTAssertEqual(password, returnPassword)
+        
+        // delete from keychain for retestability
+        var queryDelete = [String : AnyObject]()
+        queryDelete[kSecClass as String] = kSecClassGenericPassword
+        
+        SecItemDelete(queryDelete as CFDictionary)
+    }
+
```
Comment (jalwin) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: 1 2 many spaces?

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 32]
```diff
@@ -0,0 +1,69 @@
...
+        } else if passwordSize < 6  || passwordConfirmSize < 6 {
+            warningLabel.text = "Please enter a password and confirm the password"
+            return false
+        } else if passwordInput.text != passwordConfirmInput.text {
+            warningLabel.text = "Passwords must match"
+            return false
+        }
+        return true
+    }
+
```
Comment (jalwin) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: loose 1 line

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 33]
```diff
@@ -0,0 +1,69 @@
...
+                warningLabel.text = "An error has occured please try again.  error: \(status)"
+            }
+            return
+        } catch {
+            warningLabel.text = "Somthing went wrong. Please try again."
+            return
+        }
+        
+    }
+
```
Comment (jalwin) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: loose 1 line

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 34]
```diff
@@ -0,0 +1,67 @@
...
+    @IBOutlet weak var submitButton: UIButton!
+    
+    @IBAction func clickedSubmit(_ sender: Any) {
+        guard checkInput() else { return }
+
+        do {
+            let created = try KeychainActions.addAcount(username: usernameInput.text ?? "", password: passwordInput.text ?? "")
+            if created {
+                warningLabel.text = "Account created"
+                DispatchQueue.main.asyncAfter(deadline: .now() + 2.0){
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Brace spacing

Comment (prrenfrew) [reply PokemonApp/PokemonApp/CreateAccountViewController.swift]: Why are you delaying this for 2 seconds?

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 35]
```diff
@@ -0,0 +1,67 @@
...
+                    self.dismiss(animated: true, completion: nil)
+                }
+                return
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: What is the purpose of this return?

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 36]
```diff
@@ -0,0 +1,67 @@
...
+            case .noPassword:
+                warningLabel.text = "please enter a password"
+            case .unhandledError(status: -25299):
+                warningLabel.text = "That username is already being used. Please try again"
+            case .unexpectdPasswordData:
+                warningLabel.text = "An error has occured while reading the password. Please try again."
+            case .unhandledError(let status):
+                warningLabel.text = "An error has occured please try again.  error: \(status)"
+            }
+            return
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: What is the purpose of this return?

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 37]
```diff
@@ -0,0 +1,67 @@
...
+        } catch {
+            warningLabel.text = "Somthing went wrong. Please try again."
+            return
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: What is the purpose of this return?

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 38]
```diff
@@ -0,0 +1,67 @@
...
+    
+    func checkInput() -> Bool {
+        let usernameSize = usernameInput.text?.count ?? 0
+        let passwordSize = passwordInput.text?.count ?? 0
+        let passwordConfirmSize = passwordConfirmInput.text?.count ?? 0
+        
+        if usernameSize < 6{
+            warningLabel.text = "Please enter a username"
+            return false
+        } else if passwordSize < 6  || passwordConfirmSize < 6 {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Looks like business logic. Put in controller layer.

File: PokemonApp/PokemonApp/ErrorList.swift [hunk 39]
```diff
@@ -0,0 +1,15 @@
...
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import Foundation
+
+enum KeychainError: Error {
+    case noPassword
+    case unexpectdPasswordData
+    case unhandledError(status: OSStatus)
+}
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/ErrorList.swift]: Each file should have a blank line at end of file.

File: PokemonApp/PokemonApp/UserCredentials.swift [hunk 40]
```diff
@@ -0,0 +1,22 @@
...
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import Foundation
+
+struct Credentials {
+    let username: String
+    let password: String
+    
+
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/UserCredentials.swift]: Only one blank line between properties and initializer.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 41]
```diff
@@ -0,0 +1,56 @@
...
+        var query = [String : AnyObject]()
+        query[kSecClass as String] = kSecClassGenericPassword
+        query[kSecAttrAccount as String] = username as AnyObject?
+        query[kSecMatchLimit as String] = kSecMatchLimitOne
+        query[kSecReturnAttributes as String] = kCFBooleanTrue
+        query[kSecReturnData as String] = kCFBooleanTrue
+        
+        return query
+    }
+}
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: Blank line at EOF.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift (hunk 34 above)
Comment (pablo-vazquez01) [reply PokemonApp/PokemonApp/CreateAccountViewController.swift]: So that the label can show the user message saying  "Account created". Is there a better way?

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 42]
```diff
@@ -0,0 +1,29 @@
...
+
+    @IBOutlet weak var usernameInput: UITextField!
+    @IBOutlet weak var passwordInput: UITextField!
+    @IBOutlet weak var passwordConfirmInput: UITextField!
+    @IBOutlet weak var warningLabel: UILabel!
+    @IBOutlet weak var submitButton: UIButton!
+    
+    @IBAction func clickedSubmit(_ sender: Any) {
+        warningLabel.text = createAccount(username: usernameInput.text ?? "", password: passwordInput.text ?? "",confirmPassword: passwordConfirmInput.text ?? "")
+        if warningLabel.text == "Account created" {//dismiss the view after telling user the account was created
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Use a guard

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 43]
```diff
@@ -0,0 +1,29 @@
...
+            DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Remember to use weak self to avoid retain cycles

File: PokemonApp/PokemonApp/LoginController.swift [hunk 44]
```diff
@@ -0,0 +1,82 @@
...
+            return "That username is already being used. Please try again"
+        case .unexpectdPasswordData:
+            return "An error has occured while reading the password. Please try again."
+        case .unhandledError(let status):
+            return "An error has occured please try again.  error: \(status)"
+        }
+    } catch {
+        return "Somthing went wrong. Please try again."
+    }
+    return "you shouldn't be here 😱 Our trained monkeys are working hard to find out how"
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: Remove emojis.

File: PokemonApp/PokemonApp/ViewController.swift [hunk 45]
```diff
@@ -9,6 +9,11 @@
 import UIKit
 
 class ViewController: UIViewController {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/ViewController.swift]: Don't forget to make this a final class

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 46]
```diff
@@ -81,5 +81,53 @@ final class PokemonAppTests: XCTestCase {
         }
         waitForExpectations(timeout: 0.1, handler: nil)
     }
+//
+//  PokemonAppTests.swift
+//  PokemonAppTests
+//
+//  Created by Mac Consultant on 6/9/19.
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: This should only be at the time of the file. If you added unit tests, create a new file for the tests.

**Comment by jalwin** [comment]
everything lookks good.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift (hunk 42 above)
Comment (prrenfrew) [reply PokemonApp/PokemonApp/CreateAccountViewController.swift]: Still not using guard. Unresolving.

File: PokemonApp/PokemonApp/LoginController.swift [hunk 47]
```diff
@@ -0,0 +1,87 @@
...
+        
+        if let account = account {
+            query[kSecAttrAccount as String] = account as AnyObject?
+        }
+        
+        return query
+    }
+}
+
+func createAccount(username: String, password: String, confirmPassword: String) -> String {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: These look like global functions. Refrain from using global functions and variables.

File: PokemonApp/PokemonApp/LoginController.swift [hunk 48]
```diff
@@ -0,0 +1,87 @@
...
+    let usernameSize = username.count
+    let passwordSize = password.count
+    
+    if usernameSize < 6 {
+        return false
+    } else if passwordSize < 6 {
+        return false
+    } else if password != confirmPassword {
+        return false
+    }
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: All three of these if statements can be combined.

`if usernameSize < 6 || passwordSize < 6 || password != confirmPassword`

File: PokemonApp/PokemonApp/loginView.swift [hunk 49]
```diff
@@ -0,0 +1,20 @@
...
+//  loginView.swift
+//  PokemonApp
+//
+//  Created by Mac Consultant on 6/19/19.
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import UIKit
+
+final class loginView: UIViewController {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/loginView.swift]: Classes should use UpperCamelCase. Also, the name doesn't reflect what it is. I would assume loginView is a UIView, but it is a UIViewController.

File: PokemonApp/PokemonApp/loginView.swift [hunk 50]
```diff
@@ -0,0 +1,20 @@
...
+
+    @IBOutlet weak var usernameTextfield: UITextField!
+    @IBOutlet weak var passwordTextfield: UITextField!
+    @IBOutlet weak var loginButton: UIButton!
+    @IBOutlet weak var newAccountButton: UIButton!
+    @IBOutlet weak var warningLabel: UILabel!
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/loginView.swift]: So how are the values for this set? There is no controller or anything, and there is no setup in something like viewDidLoad().

File: PokemonApp/PokemonApp.xcodeproj/project.pbxproj [hunk 51]
```diff
@@ -219,19 +234,24 @@
...
 			buildActionMask = 2147483647;
 			files = (
+				029E74FF22B14B74004C09BC /* ErrorList.swift in Sources */,
+				029E74FD22AE030C004C09BC /* CreateAccountViewController.swift in Sources */,
 				02307EE622B1275D004C34BB /* NetworkController.swift in Sources */,
 				02307EED22B127CB004C34BB /* NetworkError.swift in Sources */,
-				04EDB51322A9A6C90036CF7A /* ViewController.swift in Sources */,
+				029E74ED22AAE503004C09BC /* LoginController.swift in Sources */,
 				04EDB51122A9A6C90036CF7A /* AppDelegate.swift in Sources */,
+				029E751E22BACDA4004C09BC /* loginView.swift in Sources */,
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp.xcodeproj/project.pbxproj]: Remember to change the name of the file too.

File: PokemonApp/PokemonApp.xcodeproj/project.pbxproj [hunk 52]
```diff
@@ -219,19 +234,24 @@
 			buildActionMask = 2147483647;
 			files = (
 				029DC0BA22B2D97E00E7079F /* URLMockSession.swift in Sources */,
+				029E751C22BAC9C6004C09BC /* PokemonAppTests.swift in Sources */,
 				029DC0C122B3E1FE00E7079F /* URLMockSessionInvalid.swift in Sources */,
-				02307EE122B12725004C34BB /* PokemonAppTests.swift in Sources */,
+				029E751822BAC9AC004C09BC /* newAccountTests.swift in Sources */,
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp.xcodeproj/project.pbxproj]: Name of the files should be UpperCamelCase as well.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 53]
```diff
@@ -0,0 +1,28 @@
...
+    @IBOutlet weak var passwordConfirmInput: UITextField!
+    @IBOutlet weak var warningLabel: UILabel!
+    @IBOutlet weak var submitButton: UIButton!
+    
+    @IBAction func clickedSubmit(_ sender: Any) {
+        warningLabel.text = KeychainActions.createAccount(username: usernameInput.text ?? "", password: passwordInput.text ?? "",confirmPassword: passwordConfirmInput.text ?? "")
+        guard warningLabel.text == "Account created" else { return }
+        //dismiss the view after telling user the account was created
+        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) { [weak self] in
+                self?.dismiss(animated: true, completion: nil)
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Too much indentation here.

File: PokemonApp/PokemonApp/LoginController.swift [hunk 54]
```diff
@@ -0,0 +1,84 @@
...
+        let status = SecItemAdd(query as CFDictionary, nil)
+        guard status == errSecSuccess else { throw KeychainError.unhandledError(status: status) }
+
+        return true
+    }
+
+    private static func keychainQuery(withService service: String, account: String? = nil, accessGroup: String? = nil) -> [String : AnyObject] {
+        var query = [String : AnyObject]()
+        query[kSecClass as String] = kSecClassGenericPassword
+        query[kSecAttrService as String] = service as AnyObject?
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: There a reason why you are casting to `AnyObject?` as opposed to `AnyObject`?

File: PokemonApp/PokemonApp/LoginController.swift [hunk 55]
```diff
@@ -0,0 +1,84 @@
...
+            return "Somthing went wrong. Please try again."
+        }
+        return "you shouldn't be here O.O Our trained monkeys are working hard to find out how you did that"
+    }
+
+    static func checkInput(username: String, password: String, confirmPassword: String) -> Bool {
+        let usernameSize = username.count
+        let passwordSize = password.count
+        
+        if usernameSize < 6 || passwordSize < 6 || password != confirmPassword{
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: Brace spacing.

File: PokemonApp/PokemonApp/LoginView.swift [hunk 56]
```diff
@@ -0,0 +1,21 @@
...
+//  loginView.swift
+//  PokemonApp
+//
+//  Created by Mac Consultant on 6/19/19.
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import UIKit
+
+final class LoginView: UIViewController {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginView.swift]: I would assume something that is called `LoginView` would be a UIView. It is actually a UIViewController. I would suggest something like `LoginViewController`

File: PokemonApp/PokemonApp/UserCredentials.swift [hunk 57]
```diff
@@ -0,0 +1,20 @@
...
+import Foundation
+
+struct Credentials {
+    let username: String
+    let password: String
+    
+    init(username: String, password: String) {
+        self.username = username
+        self.password = password
+    }
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/UserCredentials.swift]: This initializer is the default initializer for this struct, so there is no reason to put this here.

File: PokemonApp/PokemonAppTests/NewAccountTests.swift [hunk 58]
```diff
@@ -0,0 +1,57 @@
...
+//  Created by Mac Consultant on 6/9/19.
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+
+import XCTest
+@testable import PokemonApp
+
+final class NewAccountTests: XCTestCase {
+
+    private var username = "tester4"
+    private var password = "testerspassword3"
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/NewAccountTests.swift]: Do these ever change? If not, change them to lets.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 59]
```diff
@@ -81,5 +81,6 @@ final class PokemonAppTests: XCTestCase {
         }
         waitForExpectations(timeout: 0.1, handler: nil)
     }
+
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: Why was this line added?

Comment (pablo-vazquez01) [reply PokemonApp/PokemonAppTests/PokemonAppTests.swift]: this was merged in from develop i did not create this

File: PokemonApp/PokemonApp/LoginView.swift [hunk 60]
```diff
@@ -0,0 +1,21 @@
...
+//
+//  Created by Mac Consultant on 6/19/19.
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import UIKit
+
+final class LoginView: UIView {
+
+    @IBOutlet weak var usernameTextfield: UITextField!
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginView.swift]: Are these outlets from a storyboard?

Comment (pablo-vazquez01) [reply PokemonApp/PokemonApp/LoginView.swift]: yes

**Comment by jalwin** [comment]
looks good to me
